### PR TITLE
docs: Add a blog post about ngclient design

### PR DESCRIPTION
**Pull request by jku**
Try to explain some decisions made in ngclient (and Metadata API: I basically considered Metadata API an implementation detail of the client in this text).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

---

I think it's not quite publishing ready yet: 
* _at least links to either docs, source code or both would make sense._ EDIT: I've added links to docs. (the internal TrustedMetadataSet docs are not currently published so not linked)
* We should start linking to the blog if we publish something

... but I would appreciate review: does this feel like good blog content and is something clearly missing?

link for easier reading: https://github.com/jku/python-tuf/blob/blog-ngclient-design/docs/_posts/2022-03-22-ngclient-design.md

### CODE REVIEW

File: docs/_posts/2022-03-22-ngclient-design.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+* `Updater` (250 SLOC, [docs](https://theupdateframework.readthedocs.io/en/latest/api/tuf.ngclient.updater.html)) makes decisions on what metadata should be loaded into `TrustedMetadataSet`, both from the local cache and from a remote repository. While `TrustedMetadataSet` always raises an exception if a metadata is not valid, `Updater` considers the context and handles some failures as a part of the process and some as actual errors. `Updater` also handles persisting validated metadata and targets onto local storage and provides the user-facing API
+* `FetcherInterface` (100 SLOC, [docs](https://theupdateframework.readthedocs.io/en/latest/api/tuf.ngclient.fetcher.html)) is the abstract file downloader. By default, a Requests-based implementation is used but clients can use custom fetchers to tweak how downloads are done
```
Comment (MVrachev) [inline docs/_posts/2022-03-22-ngclient-design.md]: Maybe it's a good idea to have a quick look before releasing if the SLOC numbers are still correct.
This pr has been out for quite a time already.

File: docs/_posts/2022-03-22-ngclient-design.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+A year ago we added TUF support into pip as a prototype: this revealed some design issues that made the integration more difficult than it needed to be. As the potential pip integration is a goal for Python-TUF we wanted to smooth those rough edges.
+
+The main addition here was the `FetcherInterface`: it allows pip to keep doing all of the HTTP tweaks they have collected over the years.
+
+There were a bunch of smaller API tweaks as well: as an example, legacy Python-TUF had not anticipated downloading target files from a different host than it downloads metadata from. This is the design that PyPI uses with pypi.org and files.pythonhosted.org.
+
+# better API
+
+Since we knew we had to break API with the legacy implementation anyway, we also fixed multiple paper cuts in the API:
+ * Actual data structures are now exposed instead of dictionary "blobs"
```
Comment (MVrachev) [inline docs/_posts/2022-03-22-ngclient-design.md]: These days we have discussed how the replacement of dictionaries in favor of class attributes gives the ability to scan types with `mypy`.
Maybe you can mention that in `# Plain old software engineering`?

**Review by lukpueh — APPROVED** [review]
Great write-up, @jku! I think this can be published as is.

File: docs/_posts/2022-03-22-ngclient-design.md (hunk 1 above)
Comment (jku) [reply docs/_posts/2022-03-22-ngclient-design.md]: true but  a blog post will go stale anyway... the numbers are there to indicate where the bulk of the code is, I think that has not changed (metadata API has probably grown by 100-200 lines).

File: docs/_posts/2022-03-22-ngclient-design.md (hunk 2 above)
Comment (jku) [reply docs/_posts/2022-03-22-ngclient-design.md]: I think I'll take lukas' approval and publish this as is, but this comment is totally valid.

**Comment by jku** [comment]
Had to rebase anyway to get the CI status in order, so decided to update sloccount for metadata API as Martin suggested and bump the publish date to today.

**Pull request merged**